### PR TITLE
Add possibility to use multiple loadbalancerClasses and use `service.spec.loadBalancerClass`

### LINKED ISSUE #164: yawol should ignore services with different loadBalancerClass
When deploying multiple loadbalancer controllers / providers in the same cluster, [it is possible to set the `loadBalancerClass` of the `Service` spec](https://kubernetes.io/docs/concepts/services-networking/service/#load-balancer-class).

## Expected behavior

Yawol should ignore any services that have a `loadBalancerClass` of a different controller already set.

## Actual behavior

Yawol creates a yawollet based on the `Service`, even if it should belong to a different controller.

### PR DESCRIPTION
**Pull request by dergeberl**
yawol-cloud-controller:
- Use `service.spec.loadBalancerClass`
- Add possibility to add multiple loadBalancerClasses via flags
  - Always add "stackit.cloud/yawol" as default
- Add `--empty-classname` flag to also react on services without a class (default is true)

**Breaking Change** because if a `className` was set in the `yawol-cloud-controller` an empty class would be ignored. To have the same behavior, it is necessary to add `--empty-classname`. Most likely this is only used in dev/debug scenarios.

fix #164

### CODE REVIEW

File: cmd/yawol-cloud-controller/main.go [hunk 1]
```diff
@@ -352,3 +365,12 @@ func getInfrastructureDefaultsFromEnvOrDie() targetcontroller.InfrastructureDefa
 		InternalLB:       pointer.Bool(internalLb),
 	}
 }
+
+func (i *loadbalancerClassNames) String() string {
+	return strings.Join(*i, ",")
+}
+
+func (i *loadbalancerClassNames) Set(value string) error {
+	*i = append(*i, value)
```
Comment (timebertt) [inline cmd/yawol-cloud-controller/main.go:374]: IIUC, you need to add multiple `--classname` flags if you want to specify multiple classes.
Does `--classname=a,b` work as well?

File: cmd/yawol-cloud-controller/main.go [hunk 2]
```diff
@@ -94,9 +99,12 @@ func main() {
 		"K8s credentials for watching the Service resources.")
 	flag.StringVar(&controlKubeconfig, "control-kubeconfig", "",
 		"K8s credentials for deploying the LoadBalancer resources.")
-	flag.StringVar(&className, "classname", "",
-		"Only listen to Services with the given className. "+
-			"Default is empty and listen to all services with out className annotation")
+	flag.Var(&classNames, "classname",
+		"Only listen to Services with the given className. Can be set multiple times. "+
+			"Always listen to "+helper.DefaultLoadbalancerClass+"."+
```
Comment (timebertt) [inline cmd/yawol-cloud-controller/main.go:104]: I would use `helper.DefaultLoadbalancerClass` as the default value of this flag and not force its inclusion.
I.e., if you don't specify `--classname`, it will reconcile services with class `stackit.cloud/yawol` or no class if `--empty-classname` is set.
But if one sets `--classname=other` we shouldn't reconcile services with class `stackit.cloud/yawol`.
Otherwise, you can't have multiple yawol cloud controllers running against the same cluster.

File: controllers/yawol-cloud-controller/targetcontroller/service_controller.go [hunk 3]
```diff
@@ -57,14 +57,9 @@ func (r *ServiceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
...
 
-	className, ok := svc.Annotations[yawolv1beta1.ServiceClassName]
-	if !ok {
-		className = ""
-	}
-
 	infraDefaults := GetMergedInfrastructureDetails(r.InfrastructureDefaults, svc)
 
-	if className != r.ClassName {
+	if !helper.CheckLoadBalancerClasses(svc, r.ClassNames) {
```
Comment (timebertt) [inline controllers/yawol-cloud-controller/targetcontroller/service_controller.go:62]: Nit: let's move this up before `infraDefaults := ...`. This should be the first thing to check during recocniliation.

Comment (timebertt) [inline controllers/yawol-cloud-controller/targetcontroller/service_controller.go:62]: Can we add a predicate that filters if a Service is relevant for this controller?

File: controllers/yawol-cloud-controller/targetcontroller/service_controller_test.go [hunk 4]
```diff
@@ -547,7 +547,88 @@ var _ = Describe("Check loadbalancer reconcile", Serial, Ordered, func() {
...
+			Consistently(func() error {
+				err := k8sClient.Get(ctx, types.NamespacedName{Name: "default--class-name-service-test1", Namespace: "default"}, &lb)
+				if err != nil {
+					return client.IgnoreNotFound(err)
+				}
+				return helper.ErrInvalidClassname
+			}, time.Second*5, time.Millisecond*500).Should(Succeed())
+		})
+
+		It("create service with correct classname in spec", func() {
```
Comment (timebertt) [inline controllers/yawol-cloud-controller/targetcontroller/service_controller_test.go:592]: Do we need more test cases for the empty class name option?

File: internal/helper/service.go [hunk 5]
```diff
@@ -164,6 +164,29 @@ func GetLoadBalancerNameFromService(service *coreV1.Service) string {
 	return service.Namespace + "--" + service.Name
 }
 
+func getLoadBalancerClass(service *coreV1.Service) string {
```
Comment (timebertt) [inline internal/helper/service.go:167]: Unit tests for the new functions?

File: internal/helper/service.go [hunk 6]
```diff
@@ -164,6 +164,29 @@ func GetLoadBalancerNameFromService(service *coreV1.Service) string {
...
+	if className, ok := service.Annotations[yawolv1beta1.ServiceClassName]; ok {
+		return className
+	}
```
Comment (timebertt) [inline internal/helper/service.go:170]:
```suggestion
	if className service.Annotations[yawolv1beta1.ServiceClassName] != "" {
		return className
	}
```

File: internal/helper/service.go [hunk 7]
```diff
@@ -164,6 +164,29 @@ func GetLoadBalancerNameFromService(service *coreV1.Service) string {
...
+
+func CheckLoadBalancerClasses(service *coreV1.Service, validClasses []string) bool {
+	serviceClassName := getLoadBalancerClass(service)
+
+	for _, validClass := range validClasses {
+		if validClass == serviceClassName {
+			return true
+		}
+	}
+	return false
```
Comment (timebertt) [inline internal/helper/service.go:187]:
```suggestion
	return sets.New(validClasses...).Has(getLoadBalancerClass(service))
```

**Review by timebertt — COMMENTED** [review]
Nice PR, some nits inline :)

File: controllers/yawol-cloud-controller/targetcontroller/service_controller.go (hunk 3 above)
Comment (dergeberl) [reply controllers/yawol-cloud-controller/targetcontroller/service_controller.go:62]: No because we delete LoadBalancers if a loadbalancer exists for a services with a "wrong" class. So it is possible to change the class.

Comment (dergeberl) [reply controllers/yawol-cloud-controller/targetcontroller/service_controller.go:62]: Same as above

File: cmd/yawol-cloud-controller/main.go (hunk 1 above)
Comment (dergeberl) [reply cmd/yawol-cloud-controller/main.go:374]: Yes, if you need multiple classes than it is needed to add this flag multiple times.
Not in this case. If we need this than we also need to parse.

Comment (timebertt) [reply cmd/yawol-cloud-controller/main.go:376]: Ok, then let's keep it this way. IIRC, the https://github.com/spf13/pflag package can do this out-of-the-box

File: controllers/yawol-cloud-controller/targetcontroller/service_controller.go (hunk 3 above)
Comment (timebertt) [reply controllers/yawol-cloud-controller/targetcontroller/service_controller.go:62]: Nvm, that was a stupid comment from my side... 🙈

Comment (timebertt) [reply controllers/yawol-cloud-controller/targetcontroller/service_controller.go:62]: We should be able to filter for this as well in the predicate. This is not actually related to this PR, so let's keep it as is.

File: cmd/yawol-cloud-controller/main.go [hunk 8]
```diff
@@ -121,7 +121,9 @@ func main() {
 	opts.BindFlags(flag.CommandLine)
 	flag.Parse()
 
-	classNames = append(classNames, helper.DefaultLoadbalancerClass)
+	if len(classNames) == 0 {
+		classNames = append(classNames, helper.DefaultLoadbalancerClass)
```
Comment (timebertt) [inline cmd/yawol-cloud-controller/main.go:125]: Nit: I think, you can simply initialize the variable with `[]string{helper.DefaultLoadbalancerClass}` before defining the flag.

Comment (timebertt) [reply cmd/yawol-cloud-controller/main.go:125]: Nvm, this won't work with the `append`-based implementation of the `Value` interface

**Review by timebertt — APPROVED** [review]
